### PR TITLE
[core] Port symbol-sort-key symbol layout property

### PR DESCRIPTION
**Pull request by alexshalamov**
Port of mapbox/mapbox-gl-js#7678
Fixes: #14028

### CODE REVIEW

File: src/mbgl/programs/symbol_program.hpp [hunk 1]
```diff
@@ -298,6 +298,48 @@ class SymbolProgram : public SymbolProgramBase {
...
+              const gfx::StencilMode& stencilMode,
+              const gfx::ColorMode& colorMode,
+              const gfx::CullFaceMode& cullFaceMode,
+              const gfx::IndexBuffer& indexBuffer,
+              const Segment<AttributeList>& segment,
+              const UniformValues& uniformValues,
+              const AttributeBindings& allAttributeBindings,
+              const TextureBindings& textureBindings,
+              const std::string& layerID) {
+        static_assert(Primitive == gfx::PrimitiveTypeOf<DrawMode>::value, "incompatible draw mode");
```
Comment (astojilj) [inline src/mbgl/programs/symbol_program.hpp]: Extracting per segment draw to separate method seems like an orthogonal, unnecessary refactoring, maybe unintentionally in this commit? It was shorter (subjective): static_asert and "if (!program)" are already there in caller before the loop.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/PropertyFactory.java [hunk 2]
```diff
@@ -1785,6 +1785,26 @@
     return new LayoutPropertyValue<>("symbol-avoid-edges", value);
   }
 
+  /**
+   * Sorts features in ascending order based on this value. Features with a higher sort key will appear above features with a lower sort key wehn they overlap. Features with a lower sort key will have priority over other features when doing placement.
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/PropertyFactory.java]: Fixed in gl-js, will update spec once we close feature gap https://github.com/mapbox/mapbox-gl-js/blob/master/src/style-spec/reference/v8.json#L967

File: platform/darwin/src/MGLSymbolStyleLayer.h [hunk 3]
```diff
@@ -1018,6 +1023,23 @@ MGL_EXPORT
  */
 @property (nonatomic, null_resettable) NSExpression *symbolPlacement;
 
+/**
+ Sorts features in ascending order based on this value. Features with a higher
+ sort key will appear above features with a lower sort key wehn they overlap.
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply platform/darwin/src/MGLSymbolStyleLayer.h]: https://github.com/mapbox/mapbox-gl-js/blob/master/src/style-spec/reference/v8.json#L967

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/Property.java [hunk 4]
```diff
@@ -110,6 +110,10 @@
 
   // SYMBOL_Z_ORDER: Controls the order in which overlapping symbols in the same layer are rendered
 
+  /**
+   * If {@link SYMBOL_SORT_KEY} is set, sort based on that. Otherwise sort symbols by their position relative to the viewport.
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/Property.java]: https://github.com/mapbox/mapbox-gl-js/pull/8157

File: src/mbgl/programs/symbol_program.hpp (hunk 1 above)
Comment (astojilj) [reply src/mbgl/programs/symbol_program.hpp]: thanks for clarifying this - segment's draw used from [diff](https://github.com/mapbox/mapbox-gl-native/pull/14386/files#diff-e28696fa5b68ae9576f684ddfdbaf98cR396) - failed to load it.

File: src/mbgl/layout/symbol_layout.cpp [hunk 5]
```diff
@@ -96,6 +96,13 @@ SymbolLayout::SymbolLayout(const BucketParameters& parameters,
         return;
     }
 
+    const bool hasSymbolSortKey = !leader.layout.get<SymbolSortKey>().isUndefined();
+    const auto symbolZOrder = layout.get<SymbolZOrder>();
+    const bool sortFeaturesByKey = symbolZOrder != SymbolZOrderType::ViewportY && hasSymbolSortKey;
+    const bool zOrderByViewportY = symbolZOrder == SymbolZOrderType::ViewportY || (symbolZOrder == SymbolZOrderType::Auto && !sortFeaturesByKey);
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply src/mbgl/layout/symbol_layout.cpp]: It is addressed in `const bool zOrderByViewportY = symbolZOrder == SymbolZOrderType::ViewportY`, if `SymbolZOrderType::Source`, then, we use source z order.

File: src/mbgl/layout/symbol_layout.cpp [hunk 6]
```diff
@@ -156,7 +163,13 @@ SymbolLayout::SymbolLayout(const BucketParameters& parameters,
         }
 
         if (ft.formattedText || ft.icon) {
-            features.push_back(std::move(ft));
+            if (sortFeaturesByKey) {
+                ft.sortKey = static_cast<unsigned int>(layout.evaluate<SymbolSortKey>(zoom, ft));
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/mbgl/layout/symbol_layout.cpp]: I see this fixed in the latest patch

**Comment by chloekraw** [comment]
heads up @mapbox/maps-ios, I believe this still needs a changelog entry in the next final release. We may need to have a discussion with GL to ensure there's a clear process for adding changelog entries for new features. @zugaldia @tmpsantos
